plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3RepositoryPlugin.java

```diff
@@ -32,44 +32,131 @@
 import org.opensearch.core.xcontent.NamedXContentRegistry;
 import org.opensearch.env.Environment;
+import org.opensearch.env.NodeEnvironment;
 import org.opensearch.indices.recovery.RecoverySettings;
 import org.opensearch.plugins.Plugin;
 import org.opensearch.plugins.ReloadablePlugin;
 import org.opensearch.plugins.RepositoryPlugin;
+import org.opensearch.repositories.RepositoriesService;
 import org.opensearch.repositories.Repository;
+import org.opensearch.repositories.s3.async.AsyncExecutorBuilder;
+import org.opensearch.repositories.s3.async.AsyncTransferEventLoopGroup;
+import org.opensearch.repositories.s3.async.AsyncTransferManager;
+import org.opensearch.script.ScriptService;
+import org.opensearch.threadpool.ExecutorBuilder;
+import org.opensearch.threadpool.FixedExecutorBuilder;
+import org.opensearch.threadpool.ThreadPool;
+import org.opensearch.watcher.ResourceWatcherService;
 
 import java.io.IOException;
 import java.nio.file.Path;
+import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.Collection;
 import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
+import java.util.function.Supplier;
 
 /**
  * A plugin to add a repository type that writes to and from the AWS S3.
  */
 public class S3RepositoryPlugin extends Plugin implements RepositoryPlugin, ReloadablePlugin {
+    private static final String PRIORITY_FUTURE_COMPLETION = "priority_future_completion";
+    private static final String PRIORITY_STREAM_READER = "priority_stream_reader";
+    private static final String FUTURE_COMPLETION = "future_completion";
+    private static final String STREAM_READER = "stream_reader";
 
     protected final S3Service service;
+    private final S3AsyncService s3AsyncService;
+
     private final Path configPath;
 
+    private AsyncExecutorBuilder priorityExecutorBuilder;
+    private AsyncExecutorBuilder normalExecutorBuilder;
+
     public S3RepositoryPlugin(final Settings settings, final Path configPath) {
-        this(settings, configPath, new S3Service(configPath));
+        this(settings, configPath, new S3Service(configPath), new S3AsyncService(configPath));
+    }
+
+    @Override
+    public List<ExecutorBuilder<?>> getExecutorBuilders(Settings settings) {
+        List<ExecutorBuilder<?>> executorBuilders = new ArrayList<>();
+        executorBuilders.add(
+            new FixedExecutorBuilder(settings, PRIORITY_FUTURE_COMPLETION, priorityPoolCount(settings), 10_000, PRIORITY_FUTURE_COMPLETION)
+        );
+        executorBuilders.add(
+            new FixedExecutorBuilder(settings, PRIORITY_STREAM_READER, priorityPoolCount(settings), 10_000, PRIORITY_STREAM_READER)
+        );
+        executorBuilders.add(new FixedExecutorBuilder(settings, FUTURE_COMPLETION, normalPoolCount(settings), 10_000, FUTURE_COMPLETION));
+        executorBuilders.add(new FixedExecutorBuilder(settings, STREAM_READER, normalPoolCount(settings), 10_000, STREAM_READER));
```

Comment:
Should we use `PrioritizedOpenSearchThreadPoolExecutor` instead and avoid two separate threadpool executors?

Reply:
How do we define priority? With single pool, we need to be very careful otherwise if one starts affecting other then there can be degrading impact. What benefit do you see with single executor over multiple?